### PR TITLE
Closes issue #141 - only audit IDs in IDREFs

### PR DESCRIPTION
**Pull request by ricksbrown**
Implements the decision in #141 
@alice PTAL

### CODE REVIEW

File: src/audits/AriaRoleNotScoped.js [hunk 1]
```diff
@@ -54,7 +54,7 @@ axs.AuditRules.addRule({
         }
         // If we made it this far then no DOM ancestor has a required scope role.
         // Now we need to check if anything aria-owns this element.
-        var owners = axs.utils.getIdReferrers('aria-owns', element);  // there can only be ONE explicit owner but that's a different test
+        var owners = axs.utils.getAriaIdReferrers(element, 'aria-owns');  // there can only be ONE explicit owner but that's a different test
```
Comment (alice) [inline src/audits/AriaRoleNotScoped.js]: Why this change?

File: src/audits/DuplicateId.js [hunk 2]
```diff
@@ -25,24 +26,33 @@ axs.AuditRules.addRule({
...
-            return false;  // Is an empty ID even an ID? I think not.
-        }
-        var selector = "[id='" + id.replace(/'/g, "\\'") + "']";
+        var selector = '[id=\'' + id.replace(/'/g, '\\\'') + '\']';
         var elementsWithId = element.ownerDocument.querySelectorAll(selector);
-        return (elementsWithId.length > 1);
+        if (elementsWithId.length > 1) {
+            var visible = [];
+            for (var i = 0; i < elementsWithId.length; i++) {
+                if (!axs.utils.isElementOrAncestorHidden(elementsWithId[i])) {
```
Comment (alice) [inline src/audits/DuplicateId.js]: I'm not sure about this part. I can kind of see the intent, but I think in practice the browser literally just picks the "first" element with the given ID whether it's visible or not. I just checked and this is certainly what Chrome does: https://code.google.com/p/chromium/codesearch#chromium/src/third_party/WebKit/Source/modules/accessibility/AXNodeObject.cpp&sq=package:chromium&type=cs&l=568&rcl=1438946804

File: src/audits/AriaRoleNotScoped.js (hunk 1 above)
Comment (RickSBrown) [inline src/audits/AriaRoleNotScoped.js]: I changed the API:
1. `axs.utils.getIdReferrers` now returns both HTML and ARIA IDREFs
2. The old "just return ARIA IDREFs" function is renamed to `getAriaIdReferrers`
3. I had to reverse the arguments because arg `attributeName` is now optional `opt_attributeName` - I'm operating on the practice of always putting optional arguments last. We could leave it the other way and pass it `null` if you prefer.

File: src/audits/DuplicateId.js (hunk 2 above)
Comment (RickSBrown) [inline src/audits/DuplicateId.js]: @alice I was trying to implement your decision here of:
> - https://github.com/GoogleChrome/accessibility-developer-tools/issues/141#issuecomment-126779515 finding all the cases where an IDREF is specified and make sure that each one refers to exactly one element (unless the element is aria-hidden or not rendered).

The line you are commenting on is my attempt to do the "aria-hidden or not rendered" part but I probably misunderstood. Wait, I get it now, it makes sense - you meant if the element with the IDREFs is hidden, not the elements with the ID.

File: src/audits/AriaRoleNotScoped.js (hunk 1 above)
Comment (alice) [inline src/audits/AriaRoleNotScoped.js]: Ah, got it, thanks!

File: src/js/AccessibilityUtils.js [hunk 3]
```diff
@@ -1041,30 +1041,99 @@ axs.utils.getQuerySelectorText = function(obj) {
...
+ * @return {NodeList} The elements that refer to this element.
+ */
+axs.utils.getHtmlIdReferrers = function(element) {
+    if (!element)
+        return null;
+    var id = element.id;
+    if (!id)
+        return null;
+    id = id.replace(/'/g, "\\'");  // make it safe to use in a selector
+    var selectorTemplates = ['[contextmenu=\'{id}\']', '[itemref~=\'{id}\']', 'button[form=\'{id}\']',
```
Comment (alice) [inline src/js/AccessibilityUtils.js]: Sorry, got distracted halfway through reviewing this! Could you make these one per line? I realise it'll be a lot of lines (20 or so?), but it's pretty hard to read like this (plus it'll make it easier to add any new ones if we realise something's missing).

File: src/audits/DuplicateId.js (hunk 2 above)
Comment (alice) [inline src/audits/DuplicateId.js]: Yep, exactly - if an element <A> refers to another element <B>, but you can't see element <A>, it doesn't really matter.

File: src/js/AccessibilityUtils.js (hunk 3 above)
Comment (RickSBrown) [inline src/js/AccessibilityUtils.js]: Ok, will do.

File: src/js/AccessibilityUtils.js [hunk 4]
```diff
@@ -1041,30 +1041,113 @@ axs.utils.getQuerySelectorText = function(obj) {
...
+ */
+axs.utils.getHtmlIdReferrers = function(element) {
+    if (!element)
+        return null;
+    var id = element.id;
+    if (!id)
+        return null;
+    id = id.replace(/'/g, "\\'");  // make it safe to use in a selector
+    var selectorTemplates = [
+        '[contextmenu=\'{id}\']',
```
Comment (alice) [inline src/js/AccessibilityUtils.js]: Would be great to comment the source of this list so we can periodically check that it's up to date.

File: src/js/AccessibilityUtils.js [hunk 5]
```diff
@@ -1041,30 +1041,113 @@ axs.utils.getQuerySelectorText = function(obj) {
...
+axs.utils.getIdReferrers = function(element) {
+    var result = [];
+    var addReferrers = function() {
+        for (var i = 0; i < referrers.length; i++) {
+            result.push(referrers[i]);
+        }
+    };
+    var referrers = axs.utils.getHtmlIdReferrers(element);
+    if (referrers) {
+        addReferrers();
```
Comment (alice) [inline src/js/AccessibilityUtils.js]: s/addReferrers()/Array.prototype.push.apply(result, referrers)/ ?

File: src/audits/DuplicateId.js [hunk 6]
```diff
@@ -25,22 +26,24 @@ axs.AuditRules.addRule({
     url: 'https://github.com/GoogleChrome/accessibility-developer-tools/wiki/Audit-Rules#ax_html_02',
     severity: axs.constants.Severity.SEVERE,
     relevantElementMatcher: function(element) {
-        return axs.browserUtils.matchSelector(element, '[id]');
+        if (element.hasAttribute('id')) {
+            var referrers = axs.utils.getIdReferrers(element);
```
Comment (alice) [inline src/audits/DuplicateId.js]: This makes sense, but makes the heading confusing. Perhaps "Any ID referred to via an IDREF must be unique in the DOM"?

File: test/audits/duplicate-id-test.js [hunk 7]
```diff
@@ -24,21 +24,35 @@ test('No duplicate IDs', function() {
...
-    var element2 = fixture.appendChild(element.cloneNode());
+    fixture.appendChild(element.cloneNode());
 
     var actual = rule.run({ scope: fixture });
-    equal(actual.result, axs.constants.AuditResult.FAIL);
-    deepEqual(actual.elements, [element, element2]);
+    equal(actual.result, axs.constants.AuditResult.NA);
 });
 
 test('Single duplicate ID but it\'s in shadow DOM', function() {
```
Comment (alice) [inline test/audits/duplicate-id-test.js]: We should duplicate this test for the IDREF used case.

File: src/js/AccessibilityUtils.js (hunk 4 above)
Comment (ricksbrown) [inline src/js/AccessibilityUtils.js]: @alice sorry I got sidetracked on other work.
Sorry there is no definitive list I can find for this.

File: src/js/AccessibilityUtils.js (hunk 5 above)
Comment (ricksbrown) [inline src/js/AccessibilityUtils.js]: Love it alice, very smart.
Unfortunately PhantomJS (2.0.0) can't handle it:
`'[object NodeList]' is not a valid argument for 'Function.prototype.apply' (evaluating 'Array.prototype.push.apply')`

We'd have to convert the NodeList to an Array first which somewhat defeats the purpose of what you were going for which was a nice one-liner instead of needing a helper function.

I could change it anyway if you prefer that style, including the conversion to array.

File: test/audits/duplicate-id-test.js (hunk 7 above)
Comment (ricksbrown) [inline test/audits/duplicate-id-test.js]: Yeah good call.

File: src/audits/DuplicateId.js (hunk 6 above)
Comment (ricksbrown) [inline src/audits/DuplicateId.js]: yeah true that. i changed it.

**Comment by ricksbrown** [comment]
@alice I've addressed your feedback and this should hopefully be good to go?

**Comment by alice** [comment]
:+1: Go ahead and merge :smile:
